nemo/core/neural_factory.py

```diff
@@ -28,18 +28,210 @@
+            self.tensor_dict[name] = None
+
+    def set_tensor(self, tensor: NmTensor, value: 'torch.Tensor'):
+        """Sets the value of tensor
+
+        args:
+            tensor (NmTensor)
+            value (torch.Tensor)
+        """
+        self.tensor_dict[tensor.unique_name] = value
+
+    def check_tensor_cached(self, unique_name: str):
+        """Checks to see the tensor value has been computed in the current step yet.
+
+        args:
+            unique_name (str): The NmTensor.unique_name that we want to check for.
+
+        returns:
+            (bool) whether the tensor with unique_name has been computed yet.
+        """
+        if self.tensor_dict[unique_name] is None:
+            return False
+        return True
+
+    def get_tensor(self, name: Union[str, NmTensor], compute: bool = True):
+        """Returns the value associated with a tensor. And optionally, computes the value of the tensor if not already
+        set.
+
+        args:
+            name (str, NmTensor): The user-defined name for a tensor or the NmTensor itself.
+            compute (bool): If True and the tensor has not already been computed, there will be an attempt to create a
+                call DAG and then do a forward pass on this call DAG to compute the tensor. If False, it will return
+                None if the tensor has not been computed yet.
+                Defaults to True.
+
+        returns:
+            (torch.tensor or None) representing the computed value of the requested name. Returns None if compute is
+            False and the tensor has not been computed yet.
+        """
+        if isinstance(name, NmTensor):
+            unique_name = name.unique_name
+        else:
+            unique_name = AppState().tensor_names[name]
+        tensor_value = self.tensor_dict[unique_name]
+        if tensor_value is None and compute:
+            nmtensor = AppState().tensor_names._nmtensor_uniname_dict[unique_name]
+            callchain = topological_sort_from_leaves([nmtensor], cached_training_state=self)
+            callchain.insert(0, ())
+            self._action.nm_graph_forward_pass(callchain, self.tensor_dict)
+            tensor_value = self.tensor_dict[unique_name]
+        return tensor_value
+
+
+def topological_sort_from_leaves(leaf_nmtensors: List[NmTensor], cached_training_state: TrainingState = None):
+    """A function that accepts a list of NmTensors that need to be computed and constructs a callchain DAG that starts
+    from a datalayerNM and can be used to compute the NmTensors.
+
+    args:
+        leaf_nmtensors (List[NmTensors]): The tensors to be computed
+        cached_training_state (TrainingState): A dictionary of already computed tensors.
+            Defaults to None meaning an empty cache.
+
+    returns:
+        top_sorted_modules: the callchain DAG
+    """
+    from nemo.backends.pytorch.nm import DataLayerNM  # TODO: Replace this with a backend agnostic data layer
```

Comment:
use module.type == datalayer?

Reply:
+1 this check "use module.type == datalayer" would make it backend-agnostic

Reply:
Moved `topological_sort_from_leaves` back into backends\pytorch\actions.py and removed this line.

Reply:
why? IMHO it is better if this is backend agnostic. And backends\pytorch\actions.py  is kind of overcrowded already.

Reply:
Please see https://github.com/NVIDIA/NeMo/pull/615#discussion_r431445854